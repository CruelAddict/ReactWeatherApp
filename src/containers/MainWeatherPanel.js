import React from 'react'
import WeatherPanel from './WeatherPanel'
import MainWeatherPanelComponent from '../components/MainWeatherPanelComponent'
import {connect} from 'react-redux'
import Header from "./Header";
import Actions from "../modules/actions";

const mapStateToProps = state => ({
    geo: state.geo
});

const mapDispatchToProps = dispatch => ({
    setMainWeather: panel => dispatch(Actions.setMainWeather(panel))
});

class MainWeatherPanel extends WeatherPanel {

    constructor() {
        super();
        this.mainLocationRefresh = this.mainLocationRefresh.bind(this)
    }

    componentDidMount() {
        this.setWeather();
        this.setState(
            this.props.geo.weatherObj
        );
    }

    componentDidUpdate(prevProps) {
        let oldData = prevProps.geo.weatherObj;
        let newData = this.props.geo.weatherObj;
        console.log(newData);
        oldData !== newData &&  this.setState(newData);
    }

    setWeather() {
        this.props.setMainWeather(this);
    }

    mainLocationRefresh() {
        this.setState({
            isLoading: true
        });
        this.setWeather();
    }

    render() {
        return (
            <div>
                <Header mainLocationRefresh={this.mainLocationRefresh}/>
                <MainWeatherPanelComponent state={this.state}/>
            </div>

        )
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(MainWeatherPanel)
