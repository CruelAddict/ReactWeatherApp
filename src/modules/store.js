import {createStore, applyMiddleware} from 'redux';
import thunk from 'redux-thunk';
import reducer from './rootReducer'
import store from "../index";

const savedItems = localStorage.getItem('favorites') ? JSON.parse(localStorage.getItem('favorites')).map(value => {
        return {
            ...value,
            isLoading: true
        };
    }) : [];

const initialState = {
    favorites: {
        items: savedItems
    },
    geo: {
        permissionGranted: null,
        coordinates: null
    }
};

export default () => {
    return createStore(
        reducer,
        initialState,
        applyMiddleware(thunk)
    )
}
