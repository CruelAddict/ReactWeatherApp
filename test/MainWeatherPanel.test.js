import React from 'react';
import {shallow, render, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as Enzyme from "enzyme";
import {act} from 'react-dom/test-utils';
import sinon from 'sinon';
const MockXMLHttpRequest = require('mock-xmlhttprequest');

import store from '../src/modules/store'
import MainWeatherPanel from '../src/containers/MainWeatherPanel';
import Actions from "../src/modules/actions";

var panel;

window.fetch = sinon.stub();

const imgData = 'fakeData';

window.URL.createObjectURL = sinon.stub().returns('fakeURL');

const blobStub = () => Promise.resolve();

const mockApiResponse = (body = {}) => {
    return {
        status: 200,
        headers: { 'Content-type': 'application/json' },
        body: JSON.stringify(body),
        blob: blobStub
    };
};

window.fetch.returns(Promise.resolve(mockApiResponse(imgData)));


Enzyme.configure({adapter: new Adapter()});

const server = MockXMLHttpRequest.newServer({
    get: ['https://api.openweathermap.org/data/2.5/weather?lat=51.51&lon=-0.13&appid=bf3565940c52aaa3383c1dbc23799bb1', {
        // status: 200 is the default
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({
            "coord": {"lon": -0.13, "lat": 51.51},
            "weather": [{"id": 803, "main": "Clouds", "description": "broken clouds", "icon": "04d"}],
            "base": "stations",
            "main": {"temp": 280.6, "pressure": 1011, "humidity": 81, "temp_min": 279.82, "temp_max": 281.48},
            "visibility": 10000,
            "wind": {"speed": 2.6, "deg": 360},
            "clouds": {"all": 75},
            "dt": 1574004909,
            "sys": {"type": 1, "id": 1412, "country": "GB", "sunrise": 1573975285, "sunset": 1574006980},
            "timezone": 0,
            "id": 2643743,
            "name": "London",
            "cod": 200
        }),
    }],
}).install();

server.get('https://api.openweathermap.org/data/2.5/weather?lat=123&lon=456&appid=bf3565940c52aaa3383c1dbc23799bb1', {
    status: 404,
    headers: {'Content-Type': 'application/json'},
    body: 'city not found'
});

describe('MainWeatherPanel', () => {

    it('Should display data when loaded', () => {
        store.dispatch(Actions.setGeolocationStatus({
            permissionGranted: true,
            coordinates: {
                latitude: 51.51,
                longitude: -0.13
            }
        })).then( () => {
            panel = mount(
                <MainWeatherPanel
                                  store={store}
                />
            );
        });

        return new Promise(resolve => {
            let waitUntilLoaded = setInterval(() => {
                panel.find('MainWeatherPanel').instance().state.isLoading === false ? resolve(waitUntilLoaded) : null;
            }, 250)
        }).then((waitUntilLoaded) => {
            clearInterval(waitUntilLoaded);
            panel.update();
            expect(panel).toMatchSnapshot();
        });
    });

    it('Should display error message if server does not respond', () => {
        store.dispatch(Actions.resetGeoStore( {
                permissionGranted: true,
                coordinates: { latitude: 123, longitude: 456 }
            }
        )).then( () => {
            panel = mount(
                <MainWeatherPanel
                    store={store}
                />
            );
        });

        return new Promise(resolve => {
            let waitUntilLoaded = setInterval(() => {
                panel.find('MainWeatherPanel').instance().state.isLoading === false ? resolve(waitUntilLoaded) : null;
            }, 250)
        }).then((waitUntilLoaded) => {
            clearInterval(waitUntilLoaded);
            panel.update();
            expect(panel).toMatchSnapshot();
        });
    });

    it('Should initially be loading', () => {
        window.fetch = sinon.stub().returns(setTimeout(()=> Promise.resolve(), 15000));
        store.dispatch(Actions.resetGeoStore( {
                permissionGranted: true,
                coordinates: {
                    latitude: 51.51,
                    longitude: -0.13 }
            }
        )).then( () => {
            console.log('IMPORTANT: '+ store.getState().toString());
            panel = mount(
                <MainWeatherPanel
                    store={store}
                />
            );
            expect(panel).toMatchSnapshot()
        });
    })

});
