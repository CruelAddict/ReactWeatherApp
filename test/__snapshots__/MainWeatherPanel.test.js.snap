// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainWeatherPanel Should display data when loaded 1`] = `
<Connect(MainWeatherPanel)
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MainWeatherPanel
    geo={
      Object {
        "coordinates": Object {
          "latitude": 51.51,
          "longitude": -0.13,
        },
        "permissionGranted": true,
        "weatherObj": Object {
          "city": "London",
          "coordinates": Object {
            "latitude": 51.51,
            "longitude": -0.13,
          },
          "isFailed": false,
          "isLoading": false,
          "weather_data": Object {
            "cloudiness": "broken clouds",
            "humidity": 81,
            "icon": "fakeURL",
            "pressure": 1011,
            "temperature": 7.4500000000000455,
            "wind": 2.6,
          },
        },
      }
    }
    setMainWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <div>
      <Header
        mainLocationRefresh={[Function]}
      >
        <_default
          refreshGeo={[Function]}
        >
          <div
            id="app-header"
          >
            <div
              className="header-text-container"
            >
              <h1
                className="header-text"
              >
                Погода здесь
              </h1>
            </div>
            <div
              className="button-container"
            >
              <button
                className="geo-refresh-button"
                onClick={[Function]}
              >
                Обновить геолокацию
              </button>
            </div>
          </div>
        </_default>
      </Header>
      <_default
        state={
          Object {
            "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
            "city": "London",
            "coordinates": Object {
              "latitude": 51.51,
              "longitude": -0.13,
            },
            "isFailed": false,
            "isLoading": false,
            "weather_data": Object {
              "cloudiness": "broken clouds",
              "humidity": 81,
              "icon": "fakeURL",
              "pressure": 1011,
              "temperature": 7.4500000000000455,
              "wind": 2.6,
            },
          }
        }
      >
        <div
          className="main-weather-panel-container"
        >
          <div
            className="main-weather-panel-main-data"
          >
            <div
              className="main-city-name-container"
            >
              <h2>
                London
              </h2>
            </div>
            <div
              className="icon-cell"
            >
              <div
                className="icon-container"
              >
                <img
                  alt="weather icon"
                  src="fakeURL"
                />
              </div>
            </div>
            <div
              className="temperature-container"
            >
              <span>
                7
                °С
              </span>
            </div>
          </div>
          <_default
            state={
              Object {
                "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
                "city": "London",
                "coordinates": Object {
                  "latitude": 51.51,
                  "longitude": -0.13,
                },
                "isFailed": false,
                "isLoading": false,
                "weather_data": Object {
                  "cloudiness": "broken clouds",
                  "humidity": 81,
                  "icon": "fakeURL",
                  "pressure": 1011,
                  "temperature": 7.4500000000000455,
                  "wind": 2.6,
                },
              }
            }
          >
            <div
              className="weather-value-list"
            >
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Ветер
                </span>
                <span
                  className="weather-value-value"
                >
                  2.6
                   m/s
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Облачность
                </span>
                <span
                  className="weather-value-value"
                >
                  Broken Clouds
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Давление
                </span>
                <span
                  className="weather-value-value"
                >
                  1011
                   hpa
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Влажность
                </span>
                <span
                  className="weather-value-value"
                >
                  81
                   %
                </span>
              </div>
              <div
                className="weather-value-row"
              >
                <span
                  className="weather-value-key"
                >
                  Координаты
                </span>
                <span
                  className="weather-value-value"
                >
                  [
                  -0.13
                  , 
                  51.51
                  ]
                </span>
              </div>
            </div>
          </_default>
        </div>
      </_default>
    </div>
  </MainWeatherPanel>
</Connect(MainWeatherPanel)>
`;

exports[`MainWeatherPanel Should display error message if server does not respond 1`] = `
<Connect(MainWeatherPanel)
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MainWeatherPanel
    geo={
      Object {
        "coordinates": Object {
          "latitude": 123,
          "longitude": 456,
        },
        "permissionGranted": true,
        "weatherObj": Object {
          "errorMessage": "Город не найден",
          "isFailed": true,
          "isLoading": false,
        },
      }
    }
    setMainWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <div>
      <Header
        mainLocationRefresh={[Function]}
      >
        <_default
          refreshGeo={[Function]}
        >
          <div
            id="app-header"
          >
            <div
              className="header-text-container"
            >
              <h1
                className="header-text"
              >
                Погода здесь
              </h1>
            </div>
            <div
              className="button-container"
            >
              <button
                className="geo-refresh-button"
                onClick={[Function]}
              >
                Обновить геолокацию
              </button>
            </div>
          </div>
        </_default>
      </Header>
      <_default
        state={
          Object {
            "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
            "errorMessage": "Город не найден",
            "isFailed": true,
            "isLoading": false,
          }
        }
      >
        <div
          className="loading-main"
        >
          <h4
            className="loading-title-main error-message-main"
          >
            Город не найден
          </h4>
          <span
            className="error-help-main"
          >
            Чтобы попробовать еще раз, нажмите кнопку "Обновить геолокацию"
          </span>
        </div>
      </_default>
    </div>
  </MainWeatherPanel>
</Connect(MainWeatherPanel)>
`;

exports[`MainWeatherPanel Should initially be loading 1`] = `
<Connect(MainWeatherPanel)
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MainWeatherPanel
    geo={
      Object {
        "coordinates": Object {
          "latitude": 51.51,
          "longitude": -0.13,
        },
        "permissionGranted": true,
      }
    }
    setMainWeather={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <div>
      <Header
        mainLocationRefresh={[Function]}
      >
        <_default
          refreshGeo={[Function]}
        >
          <div
            id="app-header"
          >
            <div
              className="header-text-container"
            >
              <h1
                className="header-text"
              >
                Погода здесь
              </h1>
            </div>
            <div
              className="button-container"
            >
              <button
                className="geo-refresh-button"
                onClick={[Function]}
              >
                Обновить геолокацию
              </button>
            </div>
          </div>
        </_default>
      </Header>
      <_default
        state={
          Object {
            "api_key": "bf3565940c52aaa3383c1dbc23799bb1",
            "isLoading": true,
          }
        }
      >
        <div
          className="loading-main"
        >
          <h4
            className="loading-title-main"
          >
            Подождите, данные загружаются
          </h4>
          <div
            className="spinner-container"
          >
            <i
              className="fa fa-refresh fa-spin fa-4x"
            />
          </div>
        </div>
      </_default>
    </div>
  </MainWeatherPanel>
</Connect(MainWeatherPanel)>
`;
